packages/keyring/src/keyring/Keyring.ts

```diff
@@ -0,0 +1,376 @@
+      mnemonic.confirmed = confirmed
+    }
+    return mnemonicFromStorage(mnemonic)
+  }
+
+  public removeMnemonic(id: string) {
+    const index = this.#data.mnemonics.findIndex((mnemonic) => mnemonic.id == id)
+    if (index === -1) throw new Error("Mnemonic not found")
+    this.#data.mnemonics.splice(index, 1)
+  }
+
+  async getMnemonicText(id: string, password: string): Promise<string> {
+    const mnemonic = this.#data.mnemonics.find((s) => s.id === id)
+    if (!mnemonic) throw new Error("Mnemonic not found")
+
+    const entropy = await decryptData(mnemonic.entropy, password)
+
+    return entropyToMnemonic(entropy)
+  }
+
+  public getExistingMnemonicId(mnemonic: string): string | null {
+    const entropy = mnemonicToEntropy(mnemonic)
+    const mnemonicId = oneWayHash(entropy)
+    return this.#data.mnemonics.some((s) => s.id === mnemonicId) ? mnemonicId : null
+  }
+
+  public getAccounts(): Account[] {
+    return this.#data.accounts.map(accountFromStorage)
+  }
+
+  public getAccount(address: string): Account | null {
+    const account = this.#data.accounts.find((s) => isAddressEqual(s.address, address))
+    return account ? accountFromStorage(account) : null
+  }
+
+  public updateAccount(address: string, { name, isPortfolio, genesisHash }: UpdateAccountOptions) {
+    const account = this.#data.accounts.find((s) => s.address === address)
+    if (!account) throw new Error("Account not found")
+
+    if (name) {
+      if (typeof name !== "string" || !name) throw new Error("name is required")
+      account.name = name
+    }
+    if (account.type === "watch-only" && isPortfolio !== undefined) {
+      if (typeof isPortfolio !== "boolean") throw new Error("confirmed is required")
+      account.isPortfolio = isPortfolio
+    }
+    // allow updating genesisHash only for contacts
+    if (account.type === "contact" && genesisHash) {
+      if (!isHexString(genesisHash)) throw new Error("genesisHash must be a hex string")
+      account.genesisHash = genesisHash
+    }
+
+    return accountFromStorage(account)
+  }
+
+  public removeAccount(address: string) {
+    const index = this.#data.accounts.findIndex((s) => isAddressEqual(s.address, address))
+    if (index === -1) throw new Error("Account not found")
+    this.#data.accounts.splice(index, 1)
+  }
+
+  public addAccountExternal(options: AddAccountExternalOptions): Account {
+    const address = normalizeAddress(options.address) // breaks if invalid address
+
+    if (this.getAccount(address)) throw new Error("Account already exists")
+
+    const account: AccountStorage = {
+      ...options,
+      address,
+      createdAt: Date.now(),
+    }
+
+    if (!isAccountExternal(account)) throw new Error("Invalid account type")
+
+    this.#data.accounts.push(account)
+
+    return accountFromStorage(account)
+  }
+
+  private async ensureMnemonic(options: AddAccountDeriveOptions, password: string) {
```

Comment:
I think a docstring explaining the purpose/usage of this method would be good! `ensureMnemonic` I think is unclear on its own

Reply:
e.g. when `options.type` is `existing-mnemonic`, we just return `options.mnemonicId` (so what have we ensured?)

I'm sure it makes sense, but I think it's not clear from the code alone!

Reply:
e.g. when options.type is existing-mnemonic, we just return options.mnemonicId (so what have we ensured?)

I'm sure it makes sense, but I think it's not clear from the code alone!